text/0000-untagged_union.md

```diff
@@ -0,0 +1,325 @@
+
+The `pub` keyword works on the union and on its fields, as with a struct.  The
+union and its fields default to private.  Using a private field in a union
+instantiation, field access, or pattern match produces an error.
+
+## Uninitialized unions
+
+The compiler should consider a union uninitialized if declared without an
+initializer.  However, providing a field during instantiation, or assigning to
+a field, should cause the compiler to treat the entire union as initialized.
+
+## Unions and traits
+
+A union may have trait implementations, using the same syntax as a struct.
+
+The compiler should produce an error if a union field has a type that
+implements the `Drop` trait.
+
+## Unions and undefined behavior
+
+Rust code must not use unions to invoke [undefined
+behavior](https://doc.rust-lang.org/nightly/reference.html#behavior-considered-undefined).
+In particular, Rust code must not use unions to break the pointer aliasing
+rules with raw pointers, or access a field containing a primitive type with an
+invalid value.
+
+## Union size and alignment
+
+A union must have the same size and alignment as an equivalent C union
+declaration for the target platform.  Typically, a union would have the maximum
+size of any of its fields, and the maximum alignment of any of its fields.
+Note that those maximums may come from different fields; for instance:
+
+```rust
+untagged_union U {
+    f1: u16,
+    f2: [u8; 4],
+}
+
+#[test]
+fn test() {
+    assert_eq!(std::mem::size_of<U>(), 4);
+    assert_eq!(std::mem::align_of<U>(), 2);
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Adding a new type of data structure would increase the complexity of the
+language and the compiler implementation, albeit marginally.  However, this
+change seems likely to provide a net reduction in the quantity and complexity
+of unsafe code.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Don't do anything, and leave users of FFI interfaces with unions to continue
+  writing complex platform-specific transmute code.
+- Create macros to define unions and access their fields.  However, such macros
+  make field accesses and pattern matching look more cumbersome and less
+  structure-like.  The implementation and use of such macros provides strong
+  motivation to seek a better solution, and indeed existing writers and users
+  of such macros have specifically requested native syntax in Rust.
+- Define unions without a new keyword `untagged_union`, such as via
+  `#[repr(union)] struct`.  This would avoid any possibility of breaking
+  existing code that uses the keyword, but would make declarations more
+  verbose, and introduce potential confusion with `struct` (or whatever
+  existing construct the `#[repr(union)]` attribute modifies).
+- Use a compound keyword like `unsafe union`, while not reserving `union` on
+  its own as a keyword, to avoid breaking use of `union` as an identifier.
+  Potentially more appealing syntax, if the Rust parser can support it.
```

Comment:
Interpreting "union" as both a keyword and as an identifier seems rather challenging to support, and potentially fragile for future parser changes.